## Aufgabe 1

Trenngerade: y=-2x-1

y>=-2x-1 für +1 Klassifikation

Die Perzeptrons 1-3 sind alle gleich. Nummer 4 besitzt die selbe Trenngerade aber die Klassifikationen sind vertauscht, da mit -1 multipliziert wird.



In [2]:
# Aufgabe 2
p  = (2,1,1)
v = (1,1)

def ExecutePerceptron(perceptron, values):
    result = 0
    for i in range(len(perceptron)):
        w = perceptron[i]
        if i == len(perceptron)-1:
            result += w
        else:
            result += w * values[i]
    return result

ExecutePerceptron(p,v)

4

In [ ]:
#AND
pAnd = (2, 2, -3)

#OR
pOr = (2, 2, -1)

#Komplement
pKomp = (-2, 1)

### XOR
XOR kann nicht implementiert werden, da im Perceptron nur eine einzelne Trenngerade gibt, gäbe es 2 Trenngeraden könnte man es daarstellen.

In [258]:
#Aufgabe 3
import random

#Datensatz
dataset = [(False, (random.uniform(-1,1),random.uniform(-1,1))) for _ in range(1000)]

#Gerade
g = (random.uniform(-1,1),random.uniform(-1,1),random.uniform(-1,1))

for i in range(len(dataset)):
    print(dataset[i])
    if ExecutePerceptron(g, dataset[i][1]) > 0:
        dataset[i] = (True, dataset[i][1])
dataset

(False, (-0.8414799637063926, -0.6572029851055086))
(False, (-0.14845017844229025, -0.5193885459117069))
(False, (-0.8353556861376723, 0.297665051386403))
(False, (-0.690332372287026, -0.5972704822007382))
(False, (-0.5507233095904345, 0.5979439779987179))
(False, (-0.6471852014202264, 0.9836791245473913))
(False, (0.9131288914318434, 0.7680008690411071))
(False, (-0.7681704804554481, -0.369865494087094))
(False, (0.9965003317678103, 0.19027661246236027))
(False, (-0.7447214318710569, -0.02255393778191883))
(False, (0.7358564695292251, -0.8619493801912497))
(False, (-0.748565044249184, -0.9594448601032486))
(False, (0.47208397413483905, 0.4512175157516012))
(False, (0.9882764672022133, -0.4613944351596315))
(False, (0.7033667696011994, 0.5385053018632744))
(False, (-0.919171756181486, 0.9559375531607399))
(False, (0.7048085037114353, 0.31310813544432503))
(False, (0.9968285502118643, 0.3389568860344243))
(False, (-0.21465700401539922, -0.5258055984467263))
(False, (0.3604174236240516, 

[(True, (-0.8414799637063926, -0.6572029851055086)),
 (False, (-0.14845017844229025, -0.5193885459117069)),
 (True, (-0.8353556861376723, 0.297665051386403)),
 (True, (-0.690332372287026, -0.5972704822007382)),
 (True, (-0.5507233095904345, 0.5979439779987179)),
 (True, (-0.6471852014202264, 0.9836791245473913)),
 (False, (0.9131288914318434, 0.7680008690411071)),
 (True, (-0.7681704804554481, -0.369865494087094)),
 (False, (0.9965003317678103, 0.19027661246236027)),
 (True, (-0.7447214318710569, -0.02255393778191883)),
 (False, (0.7358564695292251, -0.8619493801912497)),
 (False, (-0.748565044249184, -0.9594448601032486)),
 (False, (0.47208397413483905, 0.4512175157516012)),
 (False, (0.9882764672022133, -0.4613944351596315)),
 (False, (0.7033667696011994, 0.5385053018632744)),
 (True, (-0.919171756181486, 0.9559375531607399)),
 (False, (0.7048085037114353, 0.31310813544432503)),
 (False, (0.9968285502118643, 0.3389568860344243)),
 (False, (-0.21465700401539922, -0.5258055984467263)),

In [238]:
def TrainPerceptron(perceptron, dataset, maxIterations, learningRate):
    for generation in range(maxIterations):
        misclassified = []
        for d in dataset:
            res = True
            if ExecutePerceptron(perceptron, d[1]) < 0: res = False
            if res != d[0]:
                misclassified.append(d)
        if len(misclassified) == 0:
            return (perceptron, generation)
        
        r = random.choice(misclassified)
        pred = ExecutePerceptron(perceptron, r[1])
        if r[0]:
            error = 1 - pred
        else:
            error = -1 -pred
        
        percList = list(perceptron)
        for i in range(len(r[1])):
            percList[i] += learningRate * error * r[1][i]
        percList[-1] += learningRate * error
        perceptron = tuple(percList)
    return (perceptron, maxIterations)

print(g)
p = (0,0,0)
print(p)
print(TrainPerceptron(p,dataset,1000,1))

(-0.435030025075408, 0.8600016684242684, 0.08437578303879456)
(0, 0, 0)
((-0.5244056058031139, 1.9312563127486104, 0.25473054280896346), 2)


In [260]:
#Training
max = -1
min = -1
sum = 0
for _ in range(1000):
    value = TrainPerceptron(p,dataset,1000,0.1)[1]
    if(value > max):
        max = value
    if(value < min or min ==-1):
        min = value
    sum += value
print(f"trained 1000 times, average Steps: {sum/1000}, minimum steps: {min}, maximum steps: {max}")

trained 1000 times, average Steps: 934.368, minimum steps: 12, maximum steps: 1000


### Training
m=10
a=1
trained 1000 times, average Steps: 8.276, minimum steps: 4, maximum steps: 15

Daten varieeren je nach Datensatz

### Experimente
m=100
a=1
trained 1000 times, average Steps: 91.888, minimum steps: 4, maximum steps: 219

m=100
a=0.1
trained 1000 times, average Steps: 158.336, minimum steps: 5, maximum steps: 794

m=1000
a=1
trained 1000 times, average Steps: 572.47, minimum steps: 22, maximum steps: 1000
(maximale anzahl an schritten ist 1000)

m=1000
a=0.1
trained 1000 times, average Steps: 934.368, minimum steps: 12, maximum steps: 1000